mindyolo/utils/callback.py

```diff
@@ -308,3 +312,94 @@ def _run_eval(self, run_context: RunContext):
+
+    def on_train_end(self, run_context: RunContext):
+        """hooks to run on the end of training process"""
+        self.summary_record.close()
+        if run_context.enable_modelarts:
+            for p in os.listdir(run_context.summary_dir):
+                summary_file_path = os.path.join(run_context.summary_dir, p)
+                sync_data(summary_file_path, run_context.train_url + "/summary/" + summary_file_path.split("/")[-1])
+
+    def on_train_epoch_end(self, run_context: RunContext):
+        """hooks to run on the end of a training epoch"""
+        trainer = run_context.trainer
+        if trainer.data_sink:
+            for i in range(len(trainer.loss_item)):
+                self.summary_record.add_value("scalar", f"{trainer.loss_item_name[i]}", trainer.loss_item[i])
+            self.summary_record.add_value("scalar", f"cur_lr", Tensor(trainer.cur_lr))
+            self.summary_record.record(run_context.cur_epoch_index)
+            self.summary_record.flush()
+
+    def on_train_step_end(self, run_context: RunContext):
+        """hooks to run on the end of a training step"""
+        trainer = run_context.trainer
+        if run_context.cur_step_index % trainer.log_interval == 0:
+            for i in range(len(trainer.loss_item)):
+                self.summary_record.add_value("scalar", f"{trainer.loss_item_name[i]}", trainer.loss_item[i])
+            self.summary_record.add_value("scalar", f"cur_lr", Tensor(trainer.cur_lr))
+            self.summary_record.record(run_context.cur_step_index)
+            self.summary_record.flush()
+
+
+@CALLBACK_REGISTRY.registry_module()
+class MsProfiler(BaseCallback):
+    """
+    Callback of whether to collect profiler data at training time.
+
+    Example:
+        Case 1: Non-data sinking mode Collects performance data in the specified step interval.
+        Case 2: Data sink mode Collects performance data for a specified epoch interval.
+    """
+
+    def __init__(self, prof_range):
+        super().__init__()
+        self.prof_range = prof_range
+
+    def on_train_begin(self, run_context: RunContext):
+        """hooks to run on the beginning of training process"""
+        self.prof = Profiler(output_path=run_context.prof_dir, start_profile=False)
+
+    def on_train_end(self, run_context: RunContext):
+        """hooks to run on the end of training process"""
+        self.prof.analyse()
+        if run_context.enable_modelarts:
+            for p in os.listdir(run_context.prof_dir):
+                prof_file_path = os.path.join(run_context.prof_dir, p)
+                sync_data(prof_file_path, run_context.train_url + "/profiling_data/" + prof_file_path.split("/")[-1])
+
+    def on_train_epoch_begin(self, run_context: RunContext):
+        """hooks to run on the beginning of a training epoch"""
+        assert len(self.prof_range) == 2, (f"Expect the prof_range argument to be a two-digit list of epochs that "
```

Comment:
Expect the prof_range argument to be a two-element list indicating start and end epoch, but got {len(self.prof_range} instead